src/GafferOSL/OSLObject.cpp

```diff
@@ -410,7 +423,10 @@ void OSLObject::compute( Gaffer::ValuePlug *output, const Gaffer::Context *conte
 			return;
 		}
 
-		ConstShadingEnginePtr shadingEngine = this->shadingEngine( context );
+		// TODO - think I probably do need to pass the attributes in here, which means also sticking the
```

Comment:
I guess this is for the case where a substitution affects `ShadingEngine::needsAttribute()`? That seems quite unlikely, but not impossible of course. I suppose if we did the `lockgeom=false` thing for parameters containing substitutions, then that would take care of this without needing to pass the attributes at this point?

Reply:
Yeah, I think lockgeom=false should take care of this.

Reply:
I thought one of the reasons we'd come back to this approach where we substitute the network upfront was to avoid the lack of constant folding implied by `lockgeom=false`?

Reply:
If we do need to pass the attributes to `shadingEngine()` then perhaps it makes sense for `OSLObject::ShadingEngine()` to get the attributes itself internally (providing them as an optional output variable)? Since we want to use substitutions everywhere except `adjustBounds()`, it seems like it might be less error-prone for `shadingEngine()` to take full responsibility?

Reply:
> I thought one of the reasons we'd come back to this approach where we substitute the network upfront was to avoid the lack of constant folding implied by lockgeom=false?

I think we want to make it an exception to evaluate a ShaderNetwork with unsubstituted substitutions, but we're not evaluating here, just using needsAttribute, so this case is basically identical to adjustBounds

The proposal I gave on adjustBounds applies here too, so if using the STRING_PARAMETER_VALUE_UNKNOWN token to label parameters that shouldn't be locked is acceptable, then we don't need to pass the attributes to shadingEngine.